cranelift/codegen/src/simple_preopt.rs

```diff
@@ -477,96 +477,72 @@ enum BranchOrderKind {
-                    return;
-                };
-
-                if destination == next_block {
-                    return;
-                }
+    let (term_inst, term_dest, cond_inst, cond_dest, kind) = match pos.func.dfg.insts[inst] {
+        InstructionData::Jump {
+            opcode: Opcode::Jump,
+            destination,
+        } => {
+            let next_block = if let Some(next_block) = pos.func.layout.next_block(block) {
+                next_block
+            } else {
+                return;
+            };
 
-                let prev_inst = if let Some(prev_inst) = pos.func.layout.prev_inst(inst) {
-                    prev_inst
-                } else {
-                    return;
-                };
+            if destination.block(&pos.func.dfg.value_lists) == next_block {
+                return;
+            }
 
-                let prev_inst_data = &pos.func.dfg.insts[prev_inst];
+            let prev_inst = if let Some(prev_inst) = pos.func.layout.prev_inst(inst) {
+                prev_inst
+            } else {
+                return;
+            };
 
-                if let Some(prev_dest) = prev_inst_data.branch_destination() {
-                    if prev_dest != next_block {
+            match &pos.func.dfg.insts[prev_inst] {
+                &InstructionData::Branch {
+                    opcode,
+                    arg,
+                    destination: cond_dest,
+                } => {
+                    if cond_dest.block(&pos.func.dfg.value_lists) != next_block {
                         return;
                     }
-                } else {
-                    return;
-                }
 
-                match prev_inst_data {
-                    InstructionData::Branch {
-                        opcode,
-                        args: ref prev_args,
-                        destination: cond_dest,
-                    } => {
-                        let cond_arg = {
-                            let args = pos.func.dfg.inst_args(prev_inst);
-                            args[0]
-                        };
-
-                        let kind = match opcode {
-                            Opcode::Brz => BranchOrderKind::BrzToBrnz(cond_arg),
-                            Opcode::Brnz => BranchOrderKind::BrnzToBrz(cond_arg),
-                            _ => panic!("unexpected opcode"),
-                        };
-
-                        (
-                            inst,
-                            args.clone(),
-                            destination,
-                            prev_inst,
-                            prev_args.clone(),
-                            *cond_dest,
-                            kind,
-                        )
-                    }
-                    _ => return,
+                    let kind = match opcode {
+                        Opcode::Brz => BranchOrderKind::BrzToBrnz(arg),
+                        Opcode::Brnz => BranchOrderKind::BrnzToBrz(arg),
```

Comment:
Most (maybe all?) of this complexity should go away when the `br_if` migration is done, so I don't care whether you make this change before merging, but:

I just noticed we can delete the `BranchOrderKind` enum and deduplicate the `replace` calls below by just directly returning which opcode the `Branch` should use after this rewrite:
```suggestion
                        Opcode::Brz => Opcode::Brnz,
                        Opcode::Brnz => Opcode::Brz,
```

We could also get rid of some noise by moving the `replace` calls inside the inner match, instead of returning a tuple below and having to let-bind it above.